File: frontend/src/services/topicService.ts
import api, { ApiResponse } from './api';
import { Topic, CreateTopicRequest, UpdateTopicRequest, TopicStatus } from '../types';

export const topicService = {
  async getByEventId(eventId: string): Promise<Topic[]> {
    const response = await api.get<ApiResponse<Topic[]>>(`/topics/event/${eventId}`);
    return response.data.data || [];
  },

  async getById(id: string): Promise<Topic> {
    const response = await api.get<ApiResponse<Topic>>(`/topics/${id}`);
    if (!response.data.data) {
      throw new Error(response.data.message || 'Topic not found');
    }
    return response.data.data;
  },

  async create(data: CreateTopicRequest): Promise<Topic> {
    const response = await api.post<ApiResponse<Topic>>('/topics', data);
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to create topic');
    }
    return response.data.data;
  },

  async update(id: string, data: UpdateTopicRequest): Promise<Topic> {
    const response = await api.put<ApiResponse<Topic>>(`/topics/${id}`, data);
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to update topic');
    }
    return response.data.data;
  },

  async changeStatus(id: string, status: TopicStatus): Promise<Topic> {
    // Backend ChangeTopicStatusDto bekliyor: { status: TopicStatus }
    const response = await api.patch<ApiResponse<Topic>>(`/topics/${id}/status`, { status });
    if (!response.data.data) {
      throw new Error(response.data.message || 'Failed to change status');
    }
    return response.data.data;
  },

  async delete(id: string): Promise<void> {
    await api.delete(`/topics/${id}`);
  },
};

File: mobile/lib/services/topic_service.dart
import '../core/constants/api_constants.dart';
import '../models/models.dart';
import 'api_client.dart';

class TopicService {
  final ApiClient _apiClient = ApiClient();

  Future<List<Topic>> getTopics({String? eventId}) async {
    final queryParams = eventId != null ? {'eventId': eventId} : null;
    final response = await _apiClient.get(
      ApiConstants.topics,
      queryParameters: queryParams,
    );
    final List<dynamic> data = response.data;
    return data.map((json) => Topic.fromJson(json)).toList();
  }

  Future<List<Topic>> getTopicsByEvent(String eventId) async {
    final response = await _apiClient.get('${ApiConstants.topics}/event/$eventId');
    final List<dynamic> data = response.data;
    return data.map((json) => Topic.fromJson(json)).toList();
  }

  Future<Topic> getTopic(String id) async {
    final response = await _apiClient.get('${ApiConstants.topics}/$id');
    return Topic.fromJson(response.data);
  }

  Future<Topic> createTopic(CreateTopicRequest request) async {
    final response = await _apiClient.post(
      ApiConstants.topics,
      data: request.toJson(),
    );
    return Topic.fromJson(response.data);
  }

  Future<Topic> updateTopic(String id, UpdateTopicRequest request) async {
    final response = await _apiClient.put(
      '${ApiConstants.topics}/$id',
      data: request.toJson(),
    );
    return Topic.fromJson(response.data);
  }

  Future<void> deleteTopic(String id) async {
    await _apiClient.delete('${ApiConstants.topics}/$id');
  }

  Future<Topic> updateTopicStatus(String id, TopicStatus status) async {
    final response = await _apiClient.put(
      '${ApiConstants.topics}/$id/status',
      data: {'status': status.value},
    );
    return Topic.fromJson(response.data);
  }
}
